{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "tight-zimbabwe",
   "metadata": {},
   "outputs": [],
   "source": [
    "from primers import primers\n",
    "import primer3 as pr\n",
    "import melting as melt\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqUtils import GC\n",
    "from Bio.SeqUtils import MeltingTemp\n",
    "import numpy as np\n",
    "from tqdm import tqdm\n",
    "from itertools import product\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "id": "derived-client",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "100%|██████████| 10/10 [00:14<00:00,  1.46s/it]\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>fragment</th>\n",
       "      <th>primer-name</th>\n",
       "      <th>seq</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>cymRC-pCym.gb</td>\n",
       "      <td>(P168-RM)_cymRC-pCym.F</td>\n",
       "      <td>CGTTTATATACTAGAGTCAACGTTTGAATTTTGCATAAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>cymRC-pCym.gb</td>\n",
       "      <td>(P169-RM)_cymRC-pCym.R</td>\n",
       "      <td>CTCTAGTAATGGAGAAATAATACAAACAGACCAGATTGTCT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>e11-N-term.gb</td>\n",
       "      <td>(P170-RM)_e11-N-term.F</td>\n",
       "      <td>GGTCTGTTTGTATTATTTCTCCATTACTAGAGTCACACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>e11-N-term.gb</td>\n",
       "      <td>(P171-RM)_e11-N-term.R</td>\n",
       "      <td>CGTCAAGTTGTCATAATAAGGATCCTAATTGGTAACGAATC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>araC-pBAD.gb</td>\n",
       "      <td>(P172-RM)_araC-pBAD.F</td>\n",
       "      <td>CCAATTAGGATCCTTATTATGACAACTTGACGGCTAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>araC-pBAD.gb</td>\n",
       "      <td>(P173-RM)_araC-pBAD.R</td>\n",
       "      <td>ACATCCGGTGACAGCTACAGTAGAGAGTTGCGATAAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>e11-C-term-backbone.gb</td>\n",
       "      <td>(P174-RM)_e11-C-term-backbone.F</td>\n",
       "      <td>CGCAACTCTCTACTGTAGCTGTCACCGGATGTGCTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>e11-C-term-backbone.gb</td>\n",
       "      <td>(P175-RM)_e11-C-term-backbone.R</td>\n",
       "      <td>AAAATTCAAACGTTGACTCTAGTATATAAACGCAGAAAGGCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>e16-N-term.gb</td>\n",
       "      <td>(P176-RM)_e16-N-term.F</td>\n",
       "      <td>GGTCTGTTTGTATTATTTCTCCATTACTAGAGTCACAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>e16-N-term.gb</td>\n",
       "      <td>(P177-RM)_e16-N-term.R</td>\n",
       "      <td>CGTCAAGTTGTCATAATAAGGATCCTAATTGGTAACGAAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>e16-C-term-backbone.gb</td>\n",
       "      <td>(P178-RM)_e16-C-term-backbone.F</td>\n",
       "      <td>CGCAACTCTCTACTGTAGCTGTCACCGGATGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>e16-C-term-backbone.gb</td>\n",
       "      <td>(P179-RM)_e16-C-term-backbone.R</td>\n",
       "      <td>AAAATTCAAACGTTGACTCTAGTATATAAACGCAGAAAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>e41-C-term-backbone.gb</td>\n",
       "      <td>(P180-RM)_e41-C-term-backbone.F</td>\n",
       "      <td>CGCAACTCTCTACTGTAGCTGTCACCGGATGTGCTTTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>e41-C-term-backbone.gb</td>\n",
       "      <td>(P181-RM)_e41-C-term-backbone.R</td>\n",
       "      <td>AAAATTCAAACGTTGACTCTAGTATATAAACGCAGAAAGGCCCACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>e42-C-term-backbone.gb</td>\n",
       "      <td>(P182-RM)_e42-C-term-backbone.F</td>\n",
       "      <td>CGCAACTCTCTACTGTAGCTGTCACCGGATGTG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>pLuxB.gb</td>\n",
       "      <td>(P183-RM)_pLuxB.F</td>\n",
       "      <td>CGTTTATATACTAGAGACCTGTAGGATCGTACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>pLuxB.gb</td>\n",
       "      <td>(P184-RM)_pLuxB.R</td>\n",
       "      <td>CTCTAGTAATGGAGAATTTATTCGACTGTAACAAACCA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>e11-N-term.gb</td>\n",
       "      <td>(P185-RM)_e11-N-term.F</td>\n",
       "      <td>GTTACAGTCGAATAAATTCTCCATTACTAGAGTCACACA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>e11-N-term.gb</td>\n",
       "      <td>(P186-RM)_e11-N-term.R</td>\n",
       "      <td>ACCCAAGCGAGGCCCCTAAGGATCCTAATTGGTAACGAATC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>pSalTTC.gb</td>\n",
       "      <td>(P187-RM)_pSalTTC.F</td>\n",
       "      <td>CCAATTAGGATCCTTAGGGGCCTCGCTTGGGTTAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>pSalTTC.gb</td>\n",
       "      <td>(P188-RM)_pSalTTC.R</td>\n",
       "      <td>ACATCCGGTGACAGCTGTTAATAACGGTAACGGAGCAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>e11-C-term-backbone.gb</td>\n",
       "      <td>(P189-RM)_e11-C-term-backbone.F</td>\n",
       "      <td>CGTTACCGTTATTAACAGCTGTCACCGGATGTGCTT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>e11-C-term-backbone.gb</td>\n",
       "      <td>(P190-RM)_e11-C-term-backbone.R</td>\n",
       "      <td>GTACGATCCTACAGGTCTCTAGTATATAAACGCAGAAAGGCC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>e16-N-term.gb</td>\n",
       "      <td>(P191-RM)_e16-N-term.F</td>\n",
       "      <td>GTTACAGTCGAATAAATTCTCCATTACTAGAGTCACAC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>e16-N-term.gb</td>\n",
       "      <td>(P192-RM)_e16-N-term.R</td>\n",
       "      <td>ACCCAAGCGAGGCCCCTAAGGATCCTAATTGGTAACGAAT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>e16-C-term-backbone.gb</td>\n",
       "      <td>(P193-RM)_e16-C-term-backbone.F</td>\n",
       "      <td>CGTTACCGTTATTAACAGCTGTCACCGGATGT</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>e16-C-term-backbone.gb</td>\n",
       "      <td>(P194-RM)_e16-C-term-backbone.R</td>\n",
       "      <td>GTACGATCCTACAGGTCTCTAGTATATAAACGCAGAAAGG</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>e41-C-term-backbone.gb</td>\n",
       "      <td>(P195-RM)_e41-C-term-backbone.F</td>\n",
       "      <td>CGTTACCGTTATTAACAGCTGTCACCGGATGTGCTTTC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>e41-C-term-backbone.gb</td>\n",
       "      <td>(P196-RM)_e41-C-term-backbone.R</td>\n",
       "      <td>GTACGATCCTACAGGTCTCTAGTATATAAACGCAGAAAGGCCCACC</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>e42-C-term-backbone.gb</td>\n",
       "      <td>(P197-RM)_e42-C-term-backbone.R</td>\n",
       "      <td>GTACGATCCTACAGGTCTCTAGTATATAAACGCAGAAAG</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                  fragment                      primer-name  \\\n",
       "0            cymRC-pCym.gb           (P168-RM)_cymRC-pCym.F   \n",
       "1            cymRC-pCym.gb           (P169-RM)_cymRC-pCym.R   \n",
       "2            e11-N-term.gb           (P170-RM)_e11-N-term.F   \n",
       "3            e11-N-term.gb           (P171-RM)_e11-N-term.R   \n",
       "4             araC-pBAD.gb            (P172-RM)_araC-pBAD.F   \n",
       "5             araC-pBAD.gb            (P173-RM)_araC-pBAD.R   \n",
       "6   e11-C-term-backbone.gb  (P174-RM)_e11-C-term-backbone.F   \n",
       "7   e11-C-term-backbone.gb  (P175-RM)_e11-C-term-backbone.R   \n",
       "8            e16-N-term.gb           (P176-RM)_e16-N-term.F   \n",
       "9            e16-N-term.gb           (P177-RM)_e16-N-term.R   \n",
       "10  e16-C-term-backbone.gb  (P178-RM)_e16-C-term-backbone.F   \n",
       "11  e16-C-term-backbone.gb  (P179-RM)_e16-C-term-backbone.R   \n",
       "12  e41-C-term-backbone.gb  (P180-RM)_e41-C-term-backbone.F   \n",
       "13  e41-C-term-backbone.gb  (P181-RM)_e41-C-term-backbone.R   \n",
       "14  e42-C-term-backbone.gb  (P182-RM)_e42-C-term-backbone.F   \n",
       "15                pLuxB.gb                (P183-RM)_pLuxB.F   \n",
       "16                pLuxB.gb                (P184-RM)_pLuxB.R   \n",
       "17           e11-N-term.gb           (P185-RM)_e11-N-term.F   \n",
       "18           e11-N-term.gb           (P186-RM)_e11-N-term.R   \n",
       "19              pSalTTC.gb              (P187-RM)_pSalTTC.F   \n",
       "20              pSalTTC.gb              (P188-RM)_pSalTTC.R   \n",
       "21  e11-C-term-backbone.gb  (P189-RM)_e11-C-term-backbone.F   \n",
       "22  e11-C-term-backbone.gb  (P190-RM)_e11-C-term-backbone.R   \n",
       "23           e16-N-term.gb           (P191-RM)_e16-N-term.F   \n",
       "24           e16-N-term.gb           (P192-RM)_e16-N-term.R   \n",
       "25  e16-C-term-backbone.gb  (P193-RM)_e16-C-term-backbone.F   \n",
       "26  e16-C-term-backbone.gb  (P194-RM)_e16-C-term-backbone.R   \n",
       "27  e41-C-term-backbone.gb  (P195-RM)_e41-C-term-backbone.F   \n",
       "28  e41-C-term-backbone.gb  (P196-RM)_e41-C-term-backbone.R   \n",
       "29  e42-C-term-backbone.gb  (P197-RM)_e42-C-term-backbone.R   \n",
       "\n",
       "                                               seq  \n",
       "0          CGTTTATATACTAGAGTCAACGTTTGAATTTTGCATAAC  \n",
       "1        CTCTAGTAATGGAGAAATAATACAAACAGACCAGATTGTCT  \n",
       "2          GGTCTGTTTGTATTATTTCTCCATTACTAGAGTCACACA  \n",
       "3        CGTCAAGTTGTCATAATAAGGATCCTAATTGGTAACGAATC  \n",
       "4            CCAATTAGGATCCTTATTATGACAACTTGACGGCTAC  \n",
       "5           ACATCCGGTGACAGCTACAGTAGAGAGTTGCGATAAAA  \n",
       "6             CGCAACTCTCTACTGTAGCTGTCACCGGATGTGCTT  \n",
       "7       AAAATTCAAACGTTGACTCTAGTATATAAACGCAGAAAGGCC  \n",
       "8           GGTCTGTTTGTATTATTTCTCCATTACTAGAGTCACAC  \n",
       "9         CGTCAAGTTGTCATAATAAGGATCCTAATTGGTAACGAAT  \n",
       "10                CGCAACTCTCTACTGTAGCTGTCACCGGATGT  \n",
       "11        AAAATTCAAACGTTGACTCTAGTATATAAACGCAGAAAGG  \n",
       "12          CGCAACTCTCTACTGTAGCTGTCACCGGATGTGCTTTC  \n",
       "13  AAAATTCAAACGTTGACTCTAGTATATAAACGCAGAAAGGCCCACC  \n",
       "14               CGCAACTCTCTACTGTAGCTGTCACCGGATGTG  \n",
       "15               CGTTTATATACTAGAGACCTGTAGGATCGTACA  \n",
       "16          CTCTAGTAATGGAGAATTTATTCGACTGTAACAAACCA  \n",
       "17         GTTACAGTCGAATAAATTCTCCATTACTAGAGTCACACA  \n",
       "18       ACCCAAGCGAGGCCCCTAAGGATCCTAATTGGTAACGAATC  \n",
       "19             CCAATTAGGATCCTTAGGGGCCTCGCTTGGGTTAT  \n",
       "20         ACATCCGGTGACAGCTGTTAATAACGGTAACGGAGCAAA  \n",
       "21            CGTTACCGTTATTAACAGCTGTCACCGGATGTGCTT  \n",
       "22      GTACGATCCTACAGGTCTCTAGTATATAAACGCAGAAAGGCC  \n",
       "23          GTTACAGTCGAATAAATTCTCCATTACTAGAGTCACAC  \n",
       "24        ACCCAAGCGAGGCCCCTAAGGATCCTAATTGGTAACGAAT  \n",
       "25                CGTTACCGTTATTAACAGCTGTCACCGGATGT  \n",
       "26        GTACGATCCTACAGGTCTCTAGTATATAAACGCAGAAAGG  \n",
       "27          CGTTACCGTTATTAACAGCTGTCACCGGATGTGCTTTC  \n",
       "28  GTACGATCCTACAGGTCTCTAGTATATAAACGCAGAAAGGCCCACC  \n",
       "29         GTACGATCCTACAGGTCTCTAGTATATAAACGCAGAAAG  "
      ]
     },
     "execution_count": 46,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "folder = 'datasets/genbank'\n",
    "files_bulk = [\n",
    "    ['cymRC-pCym.gb', 'e11-N-term.gb', 'araC-pBAD.gb', 'e11-C-term-backbone.gb'],\n",
    "    ['cymRC-pCym.gb', 'e16-N-term.gb', 'araC-pBAD.gb', 'e16-C-term-backbone.gb'],\n",
    "    ['cymRC-pCym.gb', 'e32-N-term.gb', 'araC-pBAD.gb', 'e32-C-term-backbone.gb'],\n",
    "    ['cymRC-pCym.gb', 'e41-N-term.gb', 'araC-pBAD.gb', 'e41-C-term-backbone.gb'],\n",
    "    ['cymRC-pCym.gb', 'e42-N-term.gb', 'araC-pBAD.gb', 'e42-C-term-backbone.gb'],\n",
    "    ['pLuxB.gb', 'e11-N-term.gb', 'pSalTTC.gb', 'e11-C-term-backbone.gb'],\n",
    "    ['pLuxB.gb', 'e16-N-term.gb', 'pSalTTC.gb', 'e16-C-term-backbone.gb'],\n",
    "    ['pLuxB.gb', 'e32-N-term.gb', 'pSalTTC.gb', 'e32-C-term-backbone.gb'],\n",
    "    ['pLuxB.gb', 'e41-N-term.gb', 'pSalTTC.gb', 'e41-C-term-backbone.gb'],\n",
    "    ['pLuxB.gb', 'e42-N-term.gb', 'pSalTTC.gb', 'e42-C-term-backbone.gb']\n",
    "]\n",
    "last_num = 167\n",
    "overlap_len = 16\n",
    "\n",
    "df = pd.DataFrame()\n",
    "\n",
    "for files in tqdm(files_bulk):\n",
    "    \n",
    "    seqs = []\n",
    "    overhangs = []\n",
    "    for f in files:\n",
    "        with open('{}/{}'.format(folder, f)) as input_handle:\n",
    "            for record in SeqIO.parse(input_handle, 'genbank'):\n",
    "                seq = record.seq\n",
    "                seqs.append(seq)\n",
    "                overhangs.append((seq[:overlap_len], seq[-overlap_len:]))\n",
    "\n",
    "    fragments = [(overhangs[i-1][1], seqs[i], overhangs[(i+1)%len(seqs)][0].reverse_complement()) for i in range(len(seqs))]\n",
    "    \n",
    "    for i, f in enumerate(files):\n",
    "        fwd, rev = primers(str(fragments[i][1]), add_fwd=str(fragments[i][0]), add_rev=str(fragments[i][2]))\n",
    "        temp = pd.DataFrame([fwd, rev])\n",
    "        temp['fragment'] = f\n",
    "        df = df.append(temp)\n",
    "\n",
    "df['type'] = df['fwd'].apply(lambda x: 'F' if x else 'R')\n",
    "df = df.drop_duplicates(subset=['seq'])\n",
    "df = df.reset_index(drop=True).reset_index()\n",
    "df['primer-idx'] = df['index'].apply(lambda x: f'{(x+last_num+1):03d}')\n",
    "df['primer-name'] = '(P' + df['primer-idx'] + '-RM)_' + df['fragment'].str.split('.', expand=True)[0] + '.' + df['type']\n",
    "df[['fragment', 'primer-name', 'seq']]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "id": "apart-nerve",
   "metadata": {},
   "outputs": [],
   "source": [
    "df[['primer-name', 'seq']].to_csv('datasets/genbank/primer-list-2.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "generous-toolbox",
   "metadata": {},
   "source": [
    "### From Scratch"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "manual-western",
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_single_repetition(s):\n",
    "    \n",
    "    rep = 0\n",
    "    for i in range(len(s)-1):\n",
    "        if s[i]==s[i+1]:\n",
    "            rep += 1\n",
    "        else:\n",
    "            rep = 0\n",
    "        if rep >= 3:\n",
    "            return True\n",
    "    return False\n",
    "\n",
    "def check_double_repetition(s):\n",
    "\n",
    "    rep = 0\n",
    "    for i in range(0, len(s)-1)  :\n",
    "        if s[i:i+2]==s[i+2:i+4]:\n",
    "            rep += 1\n",
    "        else:\n",
    "            rep = 0\n",
    "        if rep >= 2:\n",
    "            True\n",
    "\n",
    "    return False\n",
    "\n",
    "def check_end_clamp(s):\n",
    "    \n",
    "    clamp = s[-5:]\n",
    "    if clamp[-1]=='G' or clamp[-1]=='C':\n",
    "        return 1\n",
    "    elif clamp[-1]=='T':\n",
    "        return -1\n",
    "    else:\n",
    "        return 0\n",
    "\n",
    "def check_gc_clamp(s):\n",
    "    \n",
    "    clamp = s[-5:]\n",
    "    bases_count = {'G': 0, 'C': 0, 'A': 0, 'T': 0}\n",
    "    for i in range(len(clamp)):\n",
    "        bases_count[clamp[i]] += 1\n",
    "    if bases_count['G'] or bases_count['C'] > 2:\n",
    "        return -1\n",
    "    return 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "governmental-mount",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder = 'datasets/genbank'\n",
    "files = ['cymrc-pcym.gb', 'e11-n-term.gb', 'arac-pbad.gb', 'e11-c-term-backbone.gb']\n",
    "seqs = []\n",
    "f_overhangs = []\n",
    "f_anneals = []\n",
    "r_overhangs = []\n",
    "r_anneals = []\n",
    "overlap_len = 16\n",
    "\n",
    "for f in files:\n",
    "    with open('{}/{}'.format(folder, f)) as input_handle:\n",
    "        for record in SeqIO.parse(input_handle, 'genbank'):\n",
    "            seq = record.seq\n",
    "            seqs.append(seq)\n",
    "            f_overhangs.append(seq[-overlap_len:])\n",
    "            r_overhangs.append(seq[:overlap_len].complement())\n",
    "            f_anneals.append([seq[:overlap] for overlap in range(18, 26)])\n",
    "            r_anneals.append([seq[-overlap:].complement() for overlap in range(18, 26)])\n",
    "num_fragments = len(seqs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "incoming-pharmacology",
   "metadata": {},
   "outputs": [],
   "source": [
    "seqs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "angry-bunch",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_overhangs = [f_overhangs[-1]] + f_overhangs[:-1]\n",
    "f_overhangs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "thirty-inclusion",
   "metadata": {},
   "outputs": [],
   "source": [
    "f_anneals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "sporting-programming",
   "metadata": {},
   "outputs": [],
   "source": [
    "r_overhangs = r_overhangs[1:] + [r_overhangs[0]]\n",
    "r_overhangs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "hazardous-election",
   "metadata": {},
   "outputs": [],
   "source": [
    "r_anneals"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "canadian-stand",
   "metadata": {},
   "outputs": [],
   "source": [
    "forward_primers = [(list(product([f_overhangs[i]], f_anneals[i]))) for i in range(num_fragments)]\n",
    "reversed_primers = [(list(product(r_anneals[i], [r_overhangs[i]]))) for i in range(num_fragments)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "olive-premiere",
   "metadata": {},
   "outputs": [],
   "source": [
    "forward_primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "exotic-lawrence",
   "metadata": {},
   "outputs": [],
   "source": [
    "reversed_primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ignored-calgary",
   "metadata": {},
   "outputs": [],
   "source": [
    "candidate_f_primers = []\n",
    "\n",
    "for forward_primer in forward_primers:\n",
    "    \n",
    "    candidate_f_primer = []\n",
    "    for primer_o, primer_a in forward_primer:\n",
    "        \n",
    "        primer_a = str(primer_a)\n",
    "        gc = GC(primer_a)\n",
    "        tm_breslaur = pr.calcTm(primer_a, tm_method=0) #breslaur\n",
    "        tm_santalucia =  pr.calcTm(primer_a, tm_method=1) #santalucia\n",
    "        \n",
    "        repeated_single = check_single_repetition(primer_a)\n",
    "        repeated_double = check_double_repetition(primer_a)\n",
    "        end_clamp_score = check_end_clamp(primer_a)\n",
    "        gc_clamp_score = check_gc_clamp(primer_a)\n",
    "        #print(primer_o, primer_a, gc, tm_santalucia)\n",
    "        \n",
    "        primer = str(primer_o) + primer_a\n",
    "        \n",
    "        hairpin = pr.calcHairpin(primer).structure_found\n",
    "        hairpin_tm = pr.calcHairpin(primer).tm\n",
    "\n",
    "        homodimer = pr.calcHomodimer(primer).structure_found\n",
    "        homodimer_tm = pr.calcHomodimer(primer).tm\n",
    "        \n",
    "        #if tm_santalucia < 40 or tm_santalucia > 60:\n",
    "        #    continue\n",
    "        #if gc < 40 or gc > 60:\n",
    "        #    continue\n",
    "        #if hairpin and (tm_santalucia - hairpin_tm) < 10:\n",
    "        #    continue\n",
    "        #if homodimer and (tm_santalucia - homodimer_tm) < 10:\n",
    "        #    continue\n",
    "        #if repeated_single or repeated_double:\n",
    "        #    continue\n",
    "        #if end_clamp_score < 0:\n",
    "        #    continue\n",
    "            \n",
    "        candidate_f_primer.append((primer, tm_santalucia, gc))\n",
    "    candidate_f_primers.append(candidate_f_primer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "limiting-tradition",
   "metadata": {},
   "outputs": [],
   "source": [
    "candidate_r_primers = []\n",
    "\n",
    "for reversed_primer in reversed_primers:\n",
    "    \n",
    "    candidate_r_primer = []\n",
    "    for primer_a, primer_o in reversed_primer:\n",
    "        \n",
    "        primer_a = str(primer_a)\n",
    "        gc = GC(primer_a)\n",
    "        tm_breslaur = pr.calcTm(primer_a, tm_method=0) #breslaur\n",
    "        tm_santalucia =  pr.calcTm(primer_a, tm_method=1) #santalucia\n",
    "        \n",
    "        repeated_single = check_single_repetition(primer_a)\n",
    "        repeated_double = check_double_repetition(primer_a)\n",
    "        end_clamp_score = check_end_clamp(primer_a)\n",
    "        gc_clamp_score = check_gc_clamp(primer_a)\n",
    "        #print(primer_o, primer_a, gc, tm_santalucia)\n",
    "        \n",
    "        primer = primer_a + str(primer_o)\n",
    "        \n",
    "        hairpin = pr.calcHairpin(primer).structure_found\n",
    "        hairpin_tm = pr.calcHairpin(primer).tm\n",
    "\n",
    "        homodimer = pr.calcHomodimer(primer).structure_found\n",
    "        homodimer_tm = pr.calcHomodimer(primer).tm\n",
    "        \n",
    "        #if tm_santalucia < 40 or tm_santalucia > 60:\n",
    "        #    continue\n",
    "        #if gc < 40 or gc > 60:\n",
    "        #    continue\n",
    "        #if hairpin and (tm_santalucia - hairpin_tm) < 10:\n",
    "        #    continue\n",
    "        #if homodimer and (tm_santalucia - homodimer_tm) < 10:\n",
    "        #    continue\n",
    "        #if repeated_single or repeated_double:\n",
    "        #    continue\n",
    "        #if end_clamp_score < 0:\n",
    "        #    continue\n",
    "            \n",
    "        candidate_r_primer.append((primer, tm_santalucia, gc))\n",
    "    candidate_r_primers.append(candidate_r_primer)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "piano-management",
   "metadata": {},
   "outputs": [],
   "source": [
    "primers_combo = [(a, b) for a, b in zip(candidate_f_primers, candidate_r_primers)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "polished-registration",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_primers = []\n",
    "for x in tqdm(range(num_fragments)):\n",
    "    final_primer = []\n",
    "    for FP in primers_combo[x][0]:\n",
    "        for RP in primers_combo[x][1]:\n",
    "            heterodimer = pr.calcHeterodimer(FP[0], RP[0]).structure_found\n",
    "            heterodimer_tm = pr.calcHeterodimer(FP[0], RP[0]).tm\n",
    "            if heterodimer and (np.min([FP[1], RP[1]]) - heterodimer_tm) >= 10:\n",
    "                final_primer.append((np.abs(FP[1] - RP[1]), FP, RP))\n",
    "    final_primers.append(sorted(final_primer)[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "hispanic-national",
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "\n",
    "to_recommend = []\n",
    "for i in range(num_fragments):\n",
    "    \n",
    "    to_recommend.append((\n",
    "        str(Seq(final_primers[i][1][0])),\n",
    "        str(Seq(final_primers[i][2][0]).complement().reverse_complement())\n",
    "    ))\n",
    "to_recommend"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "southern-victoria",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_primers[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "passing-fleet",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "lightweight-cruise",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "smaller-positive",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i in range(16, 33, 8):\n",
    "    print([a[-i:] for a in f_overhangs])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "magnetic-harmony",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "            \n",
    "shifted_overhangs = [overhangs[-1]] + overhangs[:-1]\n",
    "num_fragments = len(seqs)\n",
    "\n",
    "forward_anneals = [[a[0] for a in aa] for aa in anneals]\n",
    "forward_overhangs = [[b[1] for b in bb] for bb in shifted_overhangs]\n",
    "reversed_anneals = [[a[0].complement() for a in aa] for aa in anneals]\n",
    "reversed_overhangs = [[b[1].complement() for b in bb] for bb in shifted_overhangs]\n",
    "\n",
    "forward_primers = []\n",
    "reversed_primers = []\n",
    "for i in range(num_fragments):\n",
    "    forward_primers.append(list(product(forward_overhangs[i], forward_anneals[i])))\n",
    "    reversed_primers.append(list(product(reversed_overhangs[i], reversed_anneals[i])))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "minus-command",
   "metadata": {},
   "outputs": [],
   "source": [
    "[len(seq) for seq in seqs]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "advised-watershed",
   "metadata": {},
   "outputs": [],
   "source": [
    "forward_primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "answering-tracy",
   "metadata": {},
   "outputs": [],
   "source": [
    "reversed_primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "pursuant-absorption",
   "metadata": {},
   "outputs": [],
   "source": [
    "for i, fragment in tqdm(enumerate(fragments)):\n",
    "    \n",
    "    \n",
    "    for i in range(16, 40, 8):\n",
    "    for j in range(0, i+1, 8):\n",
    "        overhangs = [fragment[0][:j], fragment[1][:(i-j)]]\n",
    "        \n",
    "        \n",
    "        \n",
    "\n",
    "    seq = fragment[-1]\n",
    "    primers = [[], []]\n",
    "    final_primers = []\n",
    "\n",
    "    for i in range(18, 25):\n",
    "\n",
    "        candidate_primers = [str(seq[:i]), str(seq[-i:].reverse_complement())]\n",
    "        \n",
    "        for j, primer in enumerate(candidate_primers):\n",
    "\n",
    "            gc = GC(primer)\n",
    "            tm_breslaur = pr.calcTm(primer, dntp_conc=dntp, tm_method=0) #breslaur\n",
    "            tm_santalucia =  pr.calcTm(primer, dntp_conc=dntp, tm_method=1) #santalucia\n",
    "            \n",
    "            hairpin = pr.calcHairpin(primer, dntp_conc=dntp).structure_found\n",
    "            hairpin_tm = pr.calcHairpin(primer, dntp_conc=dntp).tm\n",
    "\n",
    "            homodimer = pr.calcHomodimer(primer, dntp_conc=dntp).structure_found\n",
    "            homodimer_tm = pr.calcHomodimer(primer, dntp_conc=dntp).tm\n",
    "\n",
    "            #print('Tm_Wallace', MeltingTemp.Tm_Wallace(FP))\n",
    "            #print('Tm_GC', MeltingTemp.Tm_GC(FP))\n",
    "            #print('Tm_NN', MeltingTemp.Tm_NN(FP))\n",
    "            #print('IDT', melt.temp(FP))\n",
    "\n",
    "            repeated_single = check_single_repetition(primer)\n",
    "            repeated_double = check_double_repetition(primer)\n",
    "            end_clamp_score = check_end_clamp(primer)\n",
    "            gc_clamp_score = check_gc_clamp(primer)\n",
    "\n",
    "            if tm_santalucia < 0 or tm_santalucia > 60:\n",
    "                continue\n",
    "            if gc < 40 or gc > 60:\n",
    "                continue\n",
    "            if hairpin and (tm_santalucia - hairpin_tm) < 10:\n",
    "                continue\n",
    "            if homodimer and (tm_santalucia - homodimer_tm) < 10:\n",
    "                continue\n",
    "            #if repeated_single or repeated_double:\n",
    "            #    continue\n",
    "            if end_clamp_score < 0:\n",
    "                continue\n",
    "            #if gc_clamp_score < 0:\n",
    "            #    continue\n",
    "            \n",
    "            primers[j].append((ext_primer, tm_santalucia, gc))\n",
    "            \n",
    "            for k in range(0, 40, 8):\n",
    "            \n",
    "                oh_primer = str(component[j][-k:])\n",
    "                \n",
    "            \n",
    "    print(primers)\n",
    "\n",
    "    for FP in primers[0]:\n",
    "        for RP in primers[1]:\n",
    "            heterodimer = pr.calcHeterodimer(FP[0], RP[0], dntp_conc=dntp).structure_found\n",
    "            heterodimer_tm = pr.calcHeterodimer(FP[0], RP[0], dntp_conc=dntp).tm\n",
    "            if heterodimer and (np.min([FP[1], RP[1]]) - heterodimer_tm) >= 10:\n",
    "                final_primers.append((np.abs(FP[1] - RP[1]), FP, RP))\n",
    "                \n",
    "    print(final_primers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "alike-cliff",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder = 'datasets/genbank'\n",
    "files = ['cymrc-pcym.gb', 'e11-n-term.gb', 'arac-pbad.gb', 'e11-c-term-backbone.gb']\n",
    "seqs = []\n",
    "#dntp = 0.8\n",
    "\n",
    "for i, component in tqdm(enumerate(components)):\n",
    "\n",
    "    seq = component[-1]\n",
    "    primers = [[], []]\n",
    "    final_primers = []\n",
    "\n",
    "    for i in range(18, 25):\n",
    "\n",
    "        candidate_primers = [str(seq[:i]), str(seq[-i:].reverse_complement())]\n",
    "        \n",
    "        for j, primer in enumerate(candidate_primers):\n",
    "\n",
    "            gc = GC(primer)\n",
    "            tm_breslaur = pr.calcTm(primer, dntp_conc=dntp, tm_method=0) #breslaur\n",
    "            tm_santalucia =  pr.calcTm(primer, dntp_conc=dntp, tm_method=1) #santalucia\n",
    "            \n",
    "            hairpin = pr.calcHairpin(primer, dntp_conc=dntp).structure_found\n",
    "            hairpin_tm = pr.calcHairpin(primer, dntp_conc=dntp).tm\n",
    "\n",
    "            homodimer = pr.calcHomodimer(primer, dntp_conc=dntp).structure_found\n",
    "            homodimer_tm = pr.calcHomodimer(primer, dntp_conc=dntp).tm\n",
    "\n",
    "            #print('Tm_Wallace', MeltingTemp.Tm_Wallace(FP))\n",
    "            #print('Tm_GC', MeltingTemp.Tm_GC(FP))\n",
    "            #print('Tm_NN', MeltingTemp.Tm_NN(FP))\n",
    "            #print('IDT', melt.temp(FP))\n",
    "\n",
    "            repeated_single = check_single_repetition(primer)\n",
    "            repeated_double = check_double_repetition(primer)\n",
    "            end_clamp_score = check_end_clamp(primer)\n",
    "            gc_clamp_score = check_gc_clamp(primer)\n",
    "\n",
    "            if tm_santalucia < 0 or tm_santalucia > 60:\n",
    "                continue\n",
    "            if gc < 40 or gc > 60:\n",
    "                continue\n",
    "            if hairpin and (tm_santalucia - hairpin_tm) < 10:\n",
    "                continue\n",
    "            if homodimer and (tm_santalucia - homodimer_tm) < 10:\n",
    "                continue\n",
    "            #if repeated_single or repeated_double:\n",
    "            #    continue\n",
    "            if end_clamp_score < 0:\n",
    "                continue\n",
    "            #if gc_clamp_score < 0:\n",
    "            #    continue\n",
    "            \n",
    "            primers[j].append((ext_primer, tm_santalucia, gc))\n",
    "            \n",
    "            for k in range(0, 40, 8):\n",
    "            \n",
    "                oh_primer = str(component[j][-k:])\n",
    "                \n",
    "            \n",
    "    print(primers)\n",
    "\n",
    "    for FP in primers[0]:\n",
    "        for RP in primers[1]:\n",
    "            heterodimer = pr.calcHeterodimer(FP[0], RP[0], dntp_conc=dntp).structure_found\n",
    "            heterodimer_tm = pr.calcHeterodimer(FP[0], RP[0], dntp_conc=dntp).tm\n",
    "            if heterodimer and (np.min([FP[1], RP[1]]) - heterodimer_tm) >= 10:\n",
    "                final_primers.append((np.abs(FP[1] - RP[1]), FP, RP))\n",
    "                \n",
    "    print(final_primers)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "starting-broadcast",
   "metadata": {},
   "outputs": [],
   "source": [
    "final_primers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "funny-niger",
   "metadata": {},
   "outputs": [],
   "source": [
    "sorted(final_primers)[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "sized-missile",
   "metadata": {},
   "outputs": [],
   "source": [
    "aligner = PairwiseAligner()\n",
    "aligner.mode = 'global'\n",
    "aligner.match_score = 2\n",
    "aligner.mismatch_score = -1\n",
    "alignments = aligner.align(\"TCAGACCG\", \"CCG\")\n",
    "for alignment in sorted(alignments):\n",
    "    print(\"Score = %.1f:\" % alignment.score)\n",
    "    print(alignment)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "proprietary-elimination",
   "metadata": {},
   "outputs": [],
   "source": [
    "len('CTCTAGTATATAAACGCAGAAAGGCCCAC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "excellent-occasion",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "optimum-saskatchewan",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "northern-prevention",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "numerous-burner",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "artificial-saturn",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "adverse-business",
   "metadata": {},
   "outputs": [],
   "source": [
    "folder = 'datasets/genbank'\n",
    "files = ['cymrc-pcym.gb', 'e11-n-term.gb', 'arac-pbad.gb', 'e11-c-term-backbone.gb']\n",
    "f = files[0]\n",
    "gb_file = read('{}/{}'.format(folder, f))\n",
    "gb_file"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "historical-variation",
   "metadata": {},
   "outputs": [],
   "source": [
    "gb_file.list_features()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "visible-nerve",
   "metadata": {},
   "outputs": [],
   "source": [
    "seq = gb_file.extract_feature(0)\n",
    "seq.seq"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "minute-building",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pydna.design import primer_design\n",
    "gfp_amplicon = primer_design(seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dying-hollywood",
   "metadata": {},
   "outputs": [],
   "source": [
    "gfp_amplicon.figure()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "novel-sunglasses",
   "metadata": {},
   "outputs": [],
   "source": [
    "class GC_clamps(object):\n",
    "    '''\n",
    "    Description:\n",
    "    ------------\n",
    "    Class obtains features from GC clamp such as strength and num_GC \n",
    "    NN model strength table taken from \n",
    "    Khandelwal G, Bhyravabhotla J. A Phenomenological Model for Predicting Melting Temperatures of DNA Sequences. PLOS ONE. 2010;5: e12433. doi:10.1371/journal.pone.0012433\n",
    "    '''\n",
    "    #initialize class variables\n",
    "    clamp = \"\"\n",
    "    num_GC = 0\n",
    "    strength={'GC': 13,'CC': 11,'GG': 11,'CG': 10,'AC': 10,'TC': 8,'AG': 8,'TG': 7,'GT': 10,'CT': 8,'GA': 8,'CA': 7,'AT': 7,'TT': 5,'AA': 5,'TA': 4}\n",
    "    score=0\n",
    "    #GC_Clamp class constructor\n",
    "    def __init__(self, primer):\n",
    "        self.clamp = primer[-5:]\n",
    "        for nt in range(len(self.clamp)-1):#inrement by 1 nucleotide at a time\n",
    "            NN=self.clamp[nt]+self.clamp[nt+1]#grab dinucleotides\n",
    "            self.score+=self.strength[NN]#GC clamp strength score\n",
    "        for nt in range(len(self.clamp)):\n",
    "            if self.clamp[nt]==\"G\" or self.clamp[nt]==\"C\":\n",
    "                self.num_GC+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "remarkable-summary",
   "metadata": {},
   "outputs": [],
   "source": [
    "salts={'Pfu':{\"Tris\":20.0,\"Na\":0.0,\"K\":10.0,\"Mg\":1.5,\"dNTPs\":0.8, \"MV_tot\":20.0,\"DV_tot\":1.5},\n",
    "           'LongAmp':{\"Tris\":60.0,\"Na\":0.0,\"K\":0.0,\"Mg\":2.0,\"dNTPs\":1.2, \"MV_tot\":20.0,\"DV_tot\":2.0},\n",
    "           'PhusionHF':{\"Tris\":25.0,\"Na\":0.0,\"K\":50.0,\"Mg\":1.5,\"dNTPs\":0.8, \"MV_tot\":50,\"DV_tot\":1.5},\n",
    "           'SuperMix':{\"Tris\":66.0,\"Na\":0.0,\"K\":0.0,\"Mg\":2.4,\"dNTPs\":0.88, \"MV_tot\":19.8,\"DV_tot\":2.4},\n",
    "           'Vent':{\"Tris\":20.0,\"Na\":0.0,\"K\":10.0,\"Mg\":2.0,\"dNTPs\":0.8, \"MV_tot\":20.0,\"DV_tot\":2.0},\n",
    "           'PrimeSTAR':{\"Tris\":0,\"Na\":0.0,\"K\":0,\"Mg\":1,\"dNTPs\":1.0, \"MV_tot\":50,\"DV_tot\":1.0},\n",
    "           'Custom':{\"Tris\":10,\"Na\":0.0,\"K\":50.0,\"Mg\":1.5,\"dNTPs\":0.8, \"MV_tot\":50,\"DV_tot\":1.5},\n",
    "           'Taq':{\"Tris\":0.0,\"Na\":0.0,\"K\":50.5,\"Mg\":1.5,\"dNTPs\":0.8, \"MV_tot\":50.5,\"DV_tot\":1.5},\n",
    "           'Q5':{\"Tris\":25.0,\"Na\":0.0,\"K\":50.0,\"Mg\":2.0,\"dNTPs\":0.2, \"MV_tot\":50.0,\"DV_tot\":2.0}\n",
    "            }\n",
    "def_salts={\n",
    "    'Mg':1.5, 'dNTPs':0.2, 'Na':0.5\n",
    "}"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
